feat(swap): use smart slippage

shoom3301 opened the pull request:
# Summary

Spec: https://www.notion.so/cownation/speed-Calculate-Slippage-tolerance-per-Market-e04d8c64082e452dbc82b7964baa3299

Technically the feature is super simple:
1. Fetch `/defaultSlippageTolerance` API
2. Use the value from the API response as slippage value (only if user didn't change it manually)

There are some refactoring came along with the feature. Please review it commit-by commit.
The main logic of the feature in [the last commit](https://github.com/cowprotocol/cowswap/pull/4649/commits/73b1a36e22ff3d7fdd65759e009c64a2d87557f0).

Alonth with the refactoring I changed the slippage logic a bit. Before this PR there was one slippage value for all networks, now it is per-network. It means, if you set 5% on Mainnet and then go to Arbitrum, the slippage value will be default or one was previously set on Arbitrum.

<img width="336" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/f2d2be05-e0b0-4afd-9e43-5e756a342b97">

# To Test

1. Specify a swap trade
- [ ] there should be `/defaultSlippageTolerance` request in network
2. The value should be user as slippage percent

# The feature is mocked

Since API is not connected yet, the smart slippage value is mocked to `1.33%`

elena-zh commented:
Hey @shoom3301 , great!
Some issues I've faced: 
1. 1st app opening, I have not specified a token pair and have not connected to a wallet, but I see 'the best' slippage value. IMO, we should keep a default value of 0.5% in the field until a token pair is specified. Or a default value also should come from the API request?
![image](https://github.com/cowprotocol/cowswap/assets/70885163/c8683389-036b-4e1a-8bbb-6e72d4f4b786)
2. When I pick slippage > than a max one (50%), I see wrong message 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/664fbf77-a71d-4181-b504-ad9422a7344b)
Should be this one: 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/20553982-1648-46b1-9dd5-1f61a5bd72bf)
3. AUTO slippage is selected to the ETH order (1.33 instead of 2 or 0.5 on GC and other networks)
![image](https://github.com/cowprotocol/cowswap/assets/70885163/363d481f-5e60-4c28-a3d0-9d6c32c500d2)
4. When I change ETH order slippage to 2% (Ethereum), the value is not saved in the field (2.01 is saved)
![image](https://github.com/cowprotocol/cowswap/assets/70885163/e362b5a6-f52a-44a1-ac73-df2b4adcd163)
5. When a default (auto) slippage is now selected, the slippage field is displayed on the form. WDYT about adding 'auto' word to it so it will indicate that it is auto-calculated, and if a user sets a slippage manually, do not display 'auto' label on it?
![image](https://github.com/cowprotocol/cowswap/assets/70885163/fd62763c-9d78-4289-831d-2aec4cfde6cc)
6. If a slippage is adjustable now, I think it would be great to enable the Auto option for ETH orders 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/d7e06dd1-4c7d-418d-b8f5-a885c9d1b585)
7. I cannot enter 0 into slippage field when put a cursor at the end of 1.33 and start removing numbers one by one using keyboard button
8. 0.5% slippage for a regular token pair is overwritten with 1.33 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/74a840fe-d915-426a-9810-e3ff1af36f93)

Thanks!

alfetopito commented on apps/cowswap-frontend/src/modules/swap/state/swapSlippageAtom.ts:
```diff
@@ -0,0 +1,87 @@
+type SlippageBpsPerNetwork = Record<SupportedChainId, number>
+
+const normalSwapSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>(
+  'swap-slippage-atom:v1',
+  mapSupportedNetworks(DEFAULT_SLIPPAGE_BPS)
+)
+
+const ethFlowSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>(
+  'eth-flow-slippage-atom:v1',
```
Start at v0?

alfetopito commented on apps/cowswap-frontend/src/modules/trade/state/isWrapOrUnwrapAtom.ts:
```diff
@@ -0,0 +1,13 @@
+  const { chainId } = get(walletInfoAtom)
+  const { inputCurrency, outputCurrency } = get(derivedTradeStateAtom) || {}
+
+  return getIsWrapOrUnwrap(chainId, inputCurrency?.symbol, outputCurrency?.symbol)
```
Like the other comment, we shouldn't rely on the symbol

shoom3301 commented:
@elena-zh thank you!
Fixed the issues.

5 - I think we shouldn't event display the row when slippage is "smart"
6 - this is how it works on prod now, but I agree (fixed that)
7 - this is how it works on prod, I don't think we should change this

I've removed mocks and bind the logic to the API value.
Please, be aware that 0.5% slippage is hardcoded in [https://bff.barn.cow.fi/chains/1/markets/0x6b175474e89094c44da98b954eedeac495271d0f-[…]fac5e5542a773aa44fbcfedf7c193bc2c599/defaultSlippageTolerance](https://bff.barn.cow.fi/chains/1/markets/0x6b175474e89094c44da98b954eedeac495271d0f-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/defaultSlippageTolerance), so, you will never see "smart slippage" warning, because it matches with default value. @anxolin can you please change it to 0.6% please? Just to make Elena able to test it

elena-zh commented:
Hey @shoom3301 , I will retest changes again when a new mocked value will be added. 
For now, I still can face some issues like; 
1. Possible to enter less than min slippage for ETH orders: 
![eth order less than 2](https://github.com/cowprotocol/cowswap/assets/70885163/03acbedf-a8a2-4782-83b7-1ac778fa6802)
![min slipp](https://github.com/cowprotocol/cowswap/assets/70885163/5968bfd1-5095-444a-bae9-fdc84236de31)
2. When slippage amount> max one, it should be reset to a previously used value, but not be saved in the field
![do not allow more than 50](https://github.com/cowprotocol/cowswap/assets/70885163/a5c9092b-a959-4b9d-81e8-de99945b23bc)
Thanks!

alfetopito commented on apps/cowswap-frontend/src/legacy/components/TransactionSettings/index.tsx:
```diff
@@ -1,9 +1,111 @@
+
+import useNativeCurrency from 'lib/hooks/useNativeCurrency'
+
+import { useIsSlippageModified } from '../../../modules/swap/hooks/useIsSlippageModified'
```
```suggestion
import { useIsSlippageModified } from 'modules/swap/hooks/useIsSlippageModified'
```

alfetopito commented on libs/common-const/src/bff.ts:
```diff
@@ -1 +1 @@
-export const BFF_BASE_URL = process.env.REACT_APP_BFF_BASE_URL || 'https://bff.cow.fi'
\ No newline at end of file
+export const BFF_BASE_URL = process.env.REACT_APP_BFF_BASE_URL || 'https://bff.barn.cow.fi'
```
Should it default to barn?
Need to double check the env var settings before deploying this to be safe.

shoom3301 replied:
This env var is not set by default only in localhost.
Traditionally, in localhost we use Barn environment

alfetopito commented on libs/common-const/src/common.ts:
```diff
@@ -68,22 +68,19 @@ export const ORDER_BOOK_API_UPDATE_INTERVAL = ms`30s` // 30s
-export const MINIMUM_ETH_FLOW_SLIPPAGE_BPS: Record<SupportedChainId, number> = {
+
+export const DEFAULT_ETH_FLOW_SLIPPAGE_BPS: Record<SupportedChainId, number> = {
   [SupportedChainId.MAINNET]: 200, // 2%
   [SupportedChainId.GNOSIS_CHAIN]: DEFAULT_SLIPPAGE_BPS,
   [SupportedChainId.ARBITRUM_ONE]: DEFAULT_SLIPPAGE_BPS,
   [SupportedChainId.SEPOLIA]: DEFAULT_SLIPPAGE_BPS,
 }
+
 export const MINIMUM_ETH_FLOW_SLIPPAGE: Record<SupportedChainId, Percent> = mapSupportedNetworks(
-  (chainId) => new Percent(MINIMUM_ETH_FLOW_SLIPPAGE_BPS[chainId], 10_000)
+  (chainId) => new Percent(DEFAULT_ETH_FLOW_SLIPPAGE_BPS[chainId], 10_000)
```
Sorry, I meant both of these env vars should remain `minimum`, as one is derived from the other.

shoom3301 commented:
@elena-zh thanks! Fixed

alfetopito commented on apps/cowswap-frontend/src/modules/swap/state/slippageValueAndTypeAtom.ts:
```diff
@@ -0,0 +1,72 @@
+type SlippageType = 'smart' | 'default' | 'user'
+
+const normalSwapSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>(
+  'swap-slippage-atom:v0',
+  mapSupportedNetworks(null)
+)
+
+const ethFlowSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>(
+  'eth-flow-slippage-atom:v0',
```
Sorry again, one more thing I noticed only now.

We use a mix of this and camelCase for localStorage keys.

Checking on the source, `6` are in this pattern, and `9` are using camelCase.

https://github.com/search?q=repo%3Acowprotocol%2Fcowswap+%22atom%3Av%22&type=code

I rather have a single format.
Since camelCase is the majority, let's go with that and slowly move the others to it as well.

shoom3301 replied:
Fixed

elena-zh commented:
Hey @shoom3301 , again, I've found some issues: 
1. Impossible to set 0 slippage for a regular token pair: 
![0](https://github.com/cowprotocol/cowswap/assets/70885163/b1028300-8288-44ff-bf9f-c37e91e35a14)
![return to 0 5](https://github.com/cowprotocol/cowswap/assets/70885163/93a8f69e-b9a8-4ea7-be82-73133c8e11c8)
2. For ETH orders we show 'modified' slippage on the form and on the Confirm modal even if it is set to min one. It explains why the slippage has different min values. In the PR, the label is missing on the swap form, in the Confirm modal is is replaced with a regular field
![modif](https://github.com/cowprotocol/cowswap/assets/70885163/f19824e4-e9d4-4e8b-8520-67c35c8c5098)
![modified](https://github.com/cowprotocol/cowswap/assets/70885163/e2b3b7c9-bdd8-4283-8969-9ab669844394)

Could you please fix this? 
Thanks

shoom3301 commented:
@elena-zh 
1 - fixed
2 - I can't reproduce it, could you please record a video?

elena-zh commented:
Hey @shoom3301 , great, than you!
The last issue is not fixed: ETH flow, confirm modal: '(modified)' is missing on the slippage field
![modified](https://github.com/cowprotocol/cowswap/assets/70885163/b61938a3-ca34-47df-8f4a-7d534d666911)

elena-zh reviewed (approved):
Thank you!
